edx_django_utils/monitoring/internal/backends.py

```diff
@@ -0,0 +1,177 @@
+from django.utils.module_loading import import_string
+
+log = logging.getLogger(__name__)
+
+# The newrelic package used to not be part of the requirements files
+# and so a try-import was used here. This situation is no longer true,
+# but we're still preserving that pattern until someone feels like
+# doing the work to remove it. (Should just be a major version bump
+# and communication to anyone who might be specifically removing the
+# package for some reason.)
+#
+# Ticket for just doing an unconditional import:
+# https://github.com/openedx/edx-django-utils/issues/396
+try:
+    import newrelic.agent
+except ImportError:  # pragma: no cover
+    newrelic = None  # pylint: disable=invalid-name
+
+
+class TelemetryBackend(ABC):
+    """
+    Base class for telemetry sinks.
+    """
+    @abstractmethod
+    def set_attribute(self, key, value):
+        """
+        Set a key-value attribute on a span. This might be the current
+        span or it might the root span of the process, depending on
+        the backend.
+        """
+
+    @abstractmethod
+    def record_exception(self):
+        """
+        Record the exception that is currently being handled.
+        """
+
+
+class NewRelicBackend(TelemetryBackend):
+    """
+    Send telemetry to New Relic.
+
+    https://docs.newrelic.com/docs/apm/agents/python-agent/python-agent-api/guide-using-python-agent-api/
+    """
+    def __init__(self):
+        if newrelic is None:
+            raise Exception("Could not load New Relic monitoring backend; package not present.")
+
+    def set_attribute(self, key, value):
+        # Sets attribute on the transaction, rather than the current
+        # span, matching historical behavior. There is also an
+        # `add_custom_span_attribute` that would better match
+        # OpenTelemetry's behavior, which we could try exposing
+        # through a new, more specific TelemetryBackend method.
+        #
+        # TODO: Update to newer name `add_custom_attribute`
+        # https://docs.newrelic.com/docs/apm/agents/python-agent/python-agent-api/addcustomparameter-python-agent-api/
+        newrelic.agent.add_custom_parameter(key, value)
+
+    def record_exception(self):
+        # TODO: Replace with newrelic.agent.notice_error()
+        # https://docs.newrelic.com/docs/apm/agents/python-agent/python-agent-api/recordexception-python-agent-api/
+        newrelic.agent.record_exception()
+
+
+class OpenTelemetryBackend(TelemetryBackend):
+    """
+    Send telemetry via OpenTelemetry.
+
+    Requirements to use:
+
+    - Install `opentelemetry-api` Python package
+    - Configure and initialize OpenTelemetry
+
+    API reference: https://opentelemetry-python.readthedocs.io/en/latest/
+    """
+    # pylint: disable=import-error,import-outside-toplevel
+    def __init__(self):
+        # If import fails, the backend won't be used.
+        from opentelemetry import trace
+        self.otel_trace = trace
```

Comment:
Author's note: I could add coverage for all of the backends if I add opentelemetry-api and ddtrace to the test requirements. May do so.